### ISSUE TITLE
`wpa_supplicant.conf` isn't being generated correctly when password contains `&`

### ISSUE THREAD

**Issue by leana8959** [NixOS/nixpkgs#279803]
### Describe the bug
Using the NixOS option `networking.wireless.networks.*.auth` to specify the
configuration, I noticed that if an ampersand (`&`) is present, the password
field of `wpa_supplicant.conf` isn't generated correctly.

### Steps To Reproduce
1. Use the `network.wireless.networks.*.auth` option along with the
`networking.wireless.environmentFile` option. The environment file should
contain `TEST_ME_PSK=somePassword&withAmpersand`.
2. Include the field "password=@TEST_ME_PSK@"
3. Check the content of `/run/wpa_supplicant/wpa_supplicant.conf`. The password
is decrypted to `somePassword@TEST_ME_PSK@withAmpersand`.

### Expected behavior
The password shuld be decrypted to `somePassword&withAmpersand`. The global
variable's name shouldn't show.

### Screenshots
Here are some nix code that is tested to trigger this issue instead.

```nix
networking.networkmanager.enable = lib.mkForce false;
networking.wireless = {
  enable = true;
  userControlled.enable = true;
  environmentFile = config.age.secrets.wpa_password.path;
  networks = {
    test-me = {
      authProtocols = ["WPA-EAP"];
      auth = ''
        pairwise=CCMP
        group=CCMP TKIP
        eap=PEAP
        identity="this-is-not-a-drill@null.com"
        phase2="auth=MSCHAPV2"
        password="@TEST_ME_PSK@"
        anonymous_identity="choufleur@poulet.fr"
      '';
    };
  };
};
```

The environment file:
```env
TEST_ME_PSK=somePassword&withAmpersand
```

Generated `wpa_supplicant.conf` in the aforementioned path, note how the
`password` field is erroneous.
```conf
network={
  ssid="test-me"
  key_mgmt=WPA-EAP
  pairwise=CCMP
  group=CCMP TKIP
  eap=PEAP
  identity="this-is-not-a-drill@null.com"
  phase2="auth=MSCHAPV2"
  password="somePassword@TEST_ME_PSK@withAmpersand"
  anonymous_identity="choufleur@poulet.fr"
}
```

### Additional context
A hack to get around this is to split your password in two halves, and put the &
in plain text. `@TEST_ME_PSK_L@&@TEST_ME_PSK_R@`.
Of course I would still like to see this get fixed :p

### Notify maintainers

@rnhmjoj 

Thank you for maintaining this module and reading this issue !

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.68, NixOS, 23.11 (Tapir), 23.11.20231220.d65bcea`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by rnhmjoj** [comment]
It looks like a missed a metacharacter:

> An & in the replacement text is replaced with the text that was actually
> matched. Use \& to get a literal &. (This must be typed as "\\&"; see
> https://www.gnu.org/software/gawk/manual/html_node/Gory-Details.html#Gory-Details
> for a fuller discussion of the rules for ampersands and backslashes in the
> replacement text of sub(), gsub(), and gensub().)

I don't know if escaping it as `\\&` works, or it interacts somehow with other systemd/shell/nix metacharacters. I lost counts of how many level of escaping we are in.

I hope to get rid of all of this nonsense with https://github.com/NixOS/nixpkgs/pull/180872, but next release of wpa_supplicant is taking forever.

**Comment by rnhmjoj** [comment]
Done in https://github.com/NixOS/nixpkgs/pull/305689. The fix should be in NixOS 24.05.
